### PR TITLE
Migrate AppBar to Material 3

### LINKED ISSUE #92093: Update AppBar to support Material 3
## Use case

Allow developers that are using Material 3 to have a correctly styled large/medium `AppBar`s

## Proposal

Introduce new variants to `AppBar` that allow for a larger title text that then collapses into the standard size.

See Material 3 spec at https://m3.material.io/components/top-app-bar/overview

https://user-images.githubusercontent.com/48603081/137813089-3ef62fef-d0cf-49dd-b0b3-9c3ce354268e.mp4

### PR DESCRIPTION
**Pull request by darrenaustin**
Part of: https://github.com/flutter/flutter/issues/91605

Updated the common buttons: `ElevatedButton`, `OutlinedButton`, and `TextButton` with support for [Material Design 3](https://m3.material.io/components/top-app-bar/overview).

AppBarM3_1649919991939_0AppBarM3Dark_1649920013722_0

In order to use the common buttons with the new Material 3 defaults, turn on the `useMaterial3` flag in the `ThemeData`:

```dart
  return MaterialApp(
    theme: ThemeData(useMaterial3: true),
    // ...
  );
```

In addition to the new look, the `AppBar` and `AppBarTheme` now support `surfaceTintColor` and `scrolledUnderElevation` properties to support the new elevation visuals needed for Material 3.

By default in Material 3, the app bar will no longer have a drop shadow, but instead will show changes in elevation by applying a surface tint overlay with and opacity based on the elevation on top of the background color. In addition the default elevation is 0 when there is nothing under the app bar. This changes to 3 when something is scrolled beneath the app bar.  Example of this behaviour:

AppBarScrollUnder

Fixes: #92093

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/app_bar_theme.dart [hunk 1]
```diff
@@ -320,6 +336,7 @@ class AppBarTheme with Diagnosticable {
     foregroundColor,
     elevation,
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar_theme.dart:337]:
```suggestion
    elevation,
    scrolledUnderElevation,
```

File: packages/flutter/lib/src/material/app_bar.dart [hunk 2]
```diff
@@ -383,11 +390,37 @@ class AppBar extends StatefulWidget implements PreferredSizeWidget {
...
+  ///    something scrolled underneath it.
   ///  * [shadowColor], which is the color of the shadow below the app bar.
+  ///  * [surfaceTintColor], which determines the elevation overlay that will
+  ///    be applied to the background of the app bar.
   ///  * [shape], which defines the shape of the app bar's [Material] and its
   ///    shadow.
   final double? elevation;
 
+  /// {@template flutter.material.appbar.scrolledUnderElevation}
+  /// The elevation that will be used if this App Bar has something
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar.dart:403]: nit: s/App Bar/app bar/ everywhere

**Review by guidezpl — APPROVED** [review]
LGTM! Though, there's no mention of the four types of app bar, is that something to add in a the future PR containing the examples?

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**